File: projects/LAES_turbine/LAES_turbine.py
import os
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

import barotropy as bpy

bpy.print_package_info()
bpy.set_plot_options()

# TODO: ideas to improve fitting. Doing fitting in log(density) scale
# TODO: double precission arithmetic in CFX
# TODO: custom fitting with non-negativity constraint using non-linear solver

from matplotlib import cm

# Create folder to save results
DIR_OUT = "./output"
# DIR_OUT = "./barotropic_model_20240805"
# DIR_OUT = "./barotropic_model_20240820"
if not os.path.isdir(DIR_OUT):
    os.makedirs(DIR_OUT)

# Load cases from Excel
case_table = pd.read_excel("./simulation_cases.xlsx")
case_indices = case_table["index"]
case_indices = [11, 12, 13]
# case_indices = [2]
case_table = case_table[case_table["index"].isin(case_indices)]

# Define plot settings
var_x = "s"
var_y = "T"
save_figures = True
colors = cm.magma(np.linspace(0.2, 0.8, max(len(case_indices) - 1 , 1)))  # Reds, Blues, magma

# Loop over cases
for i, (idx, row) in enumerate(case_table.iterrows()):

    # Process new row
    print(f"Processing case {i+1} of {len(case_table)}")
    print(row)
    print()
    color = "black" if i == 0 else colors[i - 1]
    print(row["tag"])
    dir_out = os.path.join(DIR_OUT, row["tag"])

    # Create barotropic model object
    fluid_name = row["fluid_name"]
    model = bpy.BarotropicModel(
        fluid_name=fluid_name,
        T_in=row["T0_in"],
        p_in=row["p0_in"],
        p_out=row["p_out"],
        efficiency=row["polytropic_efficiency"],
        calculation_type=row["calculation_type"],
        blending_onset=row["q_onset"],
        blending_width=row["q_transition"],
        # blending_width=0.05,
        HEOS_solver="hybr",
        ODE_solver="LSODA",
        ODE_tolerance=1e-9,
        polynomial_degree=8,
        polynomial_format="horner",
        output_dir=dir_out,
    )

    # Evaluate barotropic model and export polynomial expressions
    model.solve()
    model.fit_polynomials()
    model.export_expressions_fluent(output_dir=dir_out)
    model.export_expressions_cfx(output_dir=dir_out)
    for var in model.poly_fitter.variables:
        model.poly_fitter.plot_polynomial_and_error(var=var, savefig=True, showfig=True)

    # Create figure for plotting
    if i == 0:
        # Plot phase diagram
        fluid = bpy.Fluid(name=fluid_name, backend="HEOS", exceptions=True)
        fig, (ax_1, ax_2) = plt.subplots(
            1, 2, figsize=(12.0, 5.0), gridspec_kw={"wspace": 0.25}
        )
        fig.suptitle("Smooth barotropic model for expansion of nitrogen (LAES turbine case)", fontsize=14)
        ax_1.set_xlabel("Entropy (J/kg/K)\n")
        ax_1.set_ylabel("Temperature (K)")
        ax_2.set_xlabel("Pressure (Pa)\n")
        ax_2.set_ylabel("Density (kg/m$^3$)")
        # s_min, s_max = 1.3, 1.7
        # T_min, T_max = 25, 35
        # ax.set_xlim([s_min, s_max])
        # ax.set_ylim([T_min, T_max])
        ax_1 = fluid.plot_phase_diagram(
            var_x,
            var_y,
            axes=ax_1,
            plot_critical_point=True,
            plot_saturation_line=True,
            plot_spinodal_line=True,
            plot_quality_isolines=True,
            N=50,
        )

    ax_1.plot(
        model.states[var_x],
        model.states[var_y],
        # linewidth=1.0,
        # marker="o",
        # markersize=3,
        color=color,
        linewidth=1.25,
        label=rf"$q_\text{{onset}}={row['q_onset']:0.2f}$, {row['calculation_type']}",
    )

    ax_2.set_yscale("log")
    ax_2.plot(
        model.states["p"],
        model.states["rho"],
        # linewidth=1.0,
        marker="o",
        markersize=3,
        color=color,
        label=rf"$q_\text{{onset}}={row['q_onset']:0.2f}$, {row['calculation_type']}",
    )

# Add legend
ax_1.legend(loc="lower right")
ax_2.legend(loc="lower right")
# bpy.scale_graphics_x(fig, +1e-3, mode="multiply")
# bpy.scale_graphics_y(fig, -273.15, mode="add")
# fig.tight_layout(pad=1)

# Save figures
if save_figures:
    file_path = os.path.join(DIR_OUT, f"barotropic_model_{var_x}_{var_y}_diagram")
    bpy.savefig_in_formats(fig, file_path)

# Show figure
plt.show()

File: projects/LAES_turbine/LAES_turbine.m
% Clear the workspace
clear all %#ok
close all
clc

% Set plot options
set_plot_options()
save_figures = true;
defaultColors = get(groot, 'factoryAxesColorOrder');

% Create folder to save results
results_path = './data_simulations';
if not(isfolder(results_path))
    mkdir(results_path)
end

% Load experimental cases
case_table = readtable("./simulation_cases.xlsx", NumHeaderLines=0);
case_indices = case_table.index;
case_table = case_table(ismember(case_table.index, case_indices), :);

% Create the fluid
fluid_nane = "nitrogen";
fluid = FluidCoolProp_2Phase("HEOS", fluid_nane);

% Compute critical point
fluid.abstractstate.update(py.CoolProp.DmassT_INPUTS, fluid.abstractstate.rhomass_critical, fluid.abstractstate.T_critical)
T_critical = fluid.abstractstate.T;
p_critical = fluid.abstractstate.p;
s_critical = fluid.abstractstate.smass;

% Compute triple point
fluid.abstractstate.update(py.CoolProp.QT_INPUTS, 0.0, fluid.abstractstate.Ttriple)
p_triple = fluid.abstractstate.p;

% Barotropic model parameters
polynomial_order = 6;
include_metastable = false;
prop_names = ["rhomass", "viscosity", "speed_sound"];
% prop_names = ["rhomass"];

% Create figure
fig = figure(); ax = gca; hold on; box on; grid off
pbaspect([1.0, 1, 1])
ax.Title.String = {'Barotropic model'; ""};
ax.XLabel.String = {''; "$s/s_{\mathrm{crit}}$ -- Reduced entropy"};
ax.YLabel.String = {"$p/p_{\mathrm{crit}}$ -- Reduced pressure"; ''};
xtickformat('%0.2f')
ytickformat('%0.2f')
ax.XLim = [0.6, 1.1];
ax.YLim = [0, 3];


prop_x = "smass";
prop_y = "p";

% Plot phase diagram
plot_phase_diagram(prop_x, prop_y, fluid.abstractstate, ...
                   plot_critical_point=true, ...
                   plot_saturation_line=true, ...
                   plot_spinodal_line=true, ...
                   spinodal_line_color=0.5*[1,1,1], ...
                   spinodal_line_width=1.25, ...
                   spinodal_line_method='robust', ...
                   plot_quality_isolines=true, ...
                   show_in_legend=true)

% Plot barotropic model
for j = 1:height(case_table)

    % Rename variables
    index = case_table.index(j);
    T_in = case_table.T0_in(j);
    p_in = case_table.p0_in(j);
    p_out = case_table.p_out(j);

    % Define case name
    case_name = sprintf('case_%03d', index);
    fprintf('%s\n', case_name)

    % Create directory to save results
    output_folder = fullfile(results_path, case_name, "barotropic_model");
    if not(isfolder(output_folder))
        mkdir(output_folder)
    end

    % Compute properties along isentrope
    barotropic_model = create_barotropic_model(T_in, p_in, fluid, ...
                                                 include_metastable=include_metastable, ...
                                                 polynomial_order=polynomial_order, ...
                                                 properties=prop_names, ...
                                                 p_high=max(2*p_critical, 1.5*p_in), ...
                                                 p_low=1.5*p_triple, ... 
                                                 N_points=200);
    
    % Plot the isentropic expansion
    for i = 1:numel(barotropic_model.property_values)
        x = barotropic_model.property_values(i).(prop_x);
        y = barotropic_model.property_values(i).(prop_y);
        label = barotropic_model.property_values(i).label;
        if i == 1
            text(ax, x(1), y(1), 0, ['$\;\;$',num2str(case_table.index(j))], FontSize=5, HorizontalAlignment="left")
        end
        if j == 1
            plot(ax, x([1 end]), y([1 end]), Color=defaultColors(i,:), LineWidth=1.00, LineStyle='-', Marker='o', Markersize=3.5, MarkerFaceColor='w', HandleVisibility="off", DisplayName=label)
        else
            plot(ax, x([1 end]), y([1 end]), Color=defaultColors(i,:), LineWidth=1.00, LineStyle='-', Marker='o', Markersize=3.5, MarkerFaceColor='w', HandleVisibility="off", DisplayName=label)
        end
    end

    % Plot barotropic polynomials
    plot_barotropic_model(barotropic_model, case_name=case_name, output_dir=output_folder, save_figures=save_figures, show_figures=true, plot_extrapolation=true, x_scale="log")
    
    % Export fitted polynomials
    export_fluent_expressions(barotropic_model, fluid_name=fluid_nane, case_name=case_name, output_dir=output_folder)
    export_cfx_expressions(barotropic_model, fluid_name=fluid_nane, case_name=case_name, output_dir=output_folder)


end

% Display legend
legend(ax, Location="northeast")

% Rescale units
items = allchild(findall(fig, type='axes'));
for i = 1:numel(items)

    if isequal(class(items(i)), 'matlab.graphics.chart.primitive.Line') || isequal(class(items(i)), 'matlab.graphics.chart.primitive.Contour') 
        items(i).XData =  items(i).XData/s_critical;
        items(i).YData =  items(i).YData/p_critical;

    elseif isequal(class(items(i)), 'matlab.graphics.primitive.Text')
        items(i).Position(1) = items(i).Position(1)/s_critical;
        items(i).Position(2) = items(i).Position(2)/p_critical;
    end

end

% Save p-s diagram
if save_figures
    exportgraphics(fig, fullfile(results_path, 'barotropic_models_ps.png'), Resolution=500)
    savefig(fullfile(results_path, 'barotropic_models_ps.fig'))
end

